internal/k8s/reconciler/utils.go

```diff
@@ -97,12 +103,18 @@ func routeAllowedForListenerNamespaces(gatewayNS string, allowedRoutes *gw.Allow
 	case gw.NamespacesFromSame:
 		return gatewayNS == route.GetNamespace(), nil
 	case gw.NamespacesFromSelector:
-		ns, err := metav1.LabelSelectorAsSelector(namespaceSelector.Selector)
+		namespaceSelector, err := metav1.LabelSelectorAsSelector(namespaceSelector.Selector)
+		if err != nil {
+			return false, fmt.Errorf("error parsing label selector: %w", err)
+		}
+
+		// retrieve the route's namespace and determine whether selector matches
+		namespace, err := c.GetNamespace(context.TODO(), types.NamespacedName{Name: route.GetNamespace()})
 		if err != nil {
-			return false, fmt.Errorf("error parsing label selector: %v", err)
+			return false, fmt.Errorf("error retrieving namespace for route: %w", err)
 		}
 
-		return ns.Matches(toNamespaceSet(route.GetNamespace(), route.GetLabels())), nil
+		return namespaceSelector.Matches(toNamespaceSet(namespace.GetName(), namespace.GetLabels())), nil
```

Comment:
This is the meat of the PR - everything else is just to accomadate this.

Instead of attempting to apply namespace selector to route itself, retrieve the route's namespace and apply namespace selector to it.